crates/voicevox_core/src/user_dict/word.rs

```diff
@@ -0,0 +1,296 @@
+use crate::{
+    error::Error,
+    result::Result,
+    user_dict::part_of_speech_data::{
+        priority2cost, MAX_PRIORITY, MIN_PRIORITY, PART_OF_SPEECH_DETAIL,
+    },
+};
+use derive_getters::Getters;
+use once_cell::sync::Lazy;
+use regex::Regex;
+use serde::{de::Error as _, Deserialize, Serialize};
+use std::ops::RangeToInclusive;
+
+/// ユーザー辞書の単語。
+#[derive(Clone, Debug, Getters, Serialize)]
+pub struct UserDictWord {
+    /// 単語の表記。
+    pub surface: String,
+    /// 単語の読み。
+    pub pronunciation: String,
+    /// アクセント型。
+    pub accent_type: usize,
+    /// 単語の種類。
+    pub word_type: UserDictWordType,
+    /// 単語の優先度。
+    pub priority: u32,
+
+    /// モーラ数。
+    mora_count: usize,
+}
+
+impl<'de> Deserialize<'de> for UserDictWord {
+    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let raw = UserDictWord::deserialize(deserializer)?;
+        return Self::new(
+            raw.surface,
+            raw.pronunciation,
+            raw.accent_type,
+            raw.word_type,
+            raw.priority,
+        )
+        .map_err(D::Error::custom);
+
+        #[derive(Deserialize)]
+        struct UserDictWord {
+            surface: String,
+            pronunciation: String,
+            accent_type: usize,
+            word_type: UserDictWordType,
+            priority: u32,
+        }
+    }
+}
+
+#[derive(thiserror::Error, Debug, PartialEq)]
+pub enum InvalidWordError {
+    #[error("無効な発音です({1}): {0:?}")]
+    InvalidPronunciation(String, &'static str),
+    #[error("優先度は{MIN_PRIORITY}以上{MAX_PRIORITY}以下である必要があります: {0}")]
+    InvalidPriority(u32),
+    #[error("誤ったアクセント型です({1:?}の範囲から外れています): {0}")]
+    InvalidAccentType(usize, RangeToInclusive<usize>),
+}
+type InvalidWordResult<T> = std::result::Result<T, InvalidWordError>;
+
+static PRONUNCIATION_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r"^[ァ-ヴー]+$").unwrap());
+static MORA_REGEX: Lazy<Regex> = Lazy::new(|| {
+    Regex::new(concat!(
+        "(?:",
+        "[イ][ェ]|[ヴ][ャュョ]|[トド][ゥ]|[テデ][ィャュョ]|[デ][ェ]|[クグ][ヮ]|", // rule_others
+        "[キシチニヒミリギジビピ][ェャュョ]|",                                    // rule_line_i
+        "[ツフヴ][ァ]|[ウスツフヴズ][ィ]|[ウツフヴ][ェォ]|",                      // rule_line_u
+        "[ァ-ヴー]",                                                              // rule_one_mora
+        ")",
+    ))
+    .unwrap()
+});
+static SPACE_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r"\p{Z}").unwrap());
+
+impl Default for UserDictWord {
+    fn default() -> Self {
+        Self {
+            surface: "".to_string(),
+            pronunciation: "".to_string(),
+            accent_type: 0,
+            word_type: UserDictWordType::CommonNoun,
+            priority: 0,
+            mora_count: 0,
+        }
+    }
+}
+
+impl UserDictWord {
+    pub fn new(
+        surface: String,
```

Comment:
`surface`は`to_zenkaku`にかけるので、これが適切かなと思います。

```suggestion
        surface: &str, // FIXME: 呼び出し側の型を合わせる
```